### PR TITLE
Track sent and acked bytes to ensure a smooth flow of bytes not excee…

### PR DESCRIPTION
**Pull request by baldersheim**
…ding maxpending by more than

than a small byte buffer.

@hmusum and @bjorncs and @jonmv PR

### CODE REVIEW

**Comment by baldersheim** [comment]
Improves #16553

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java [hunk 1]
```diff
@@ -60,26 +61,57 @@ public void render(OutputStream output, ContentChannel networkChannel, Completio
...
+        public void send(ByteBuffer src) throws IOException {
+            try {
+                stallWhilePendingAbove(maxPending);
+            } catch (InterruptedException e) {}
+            send(src, new TrackCompletition(src.remaining()));
+        }
+
+        private void stallWhilePendingAbove(long pending) throws InterruptedException {
+            while (pendingBytes() > pending) {
+                Thread.sleep(1);
```
Comment (jonmv) [inline container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:106]: Perhaps a `java.util.concurrent.Condition` would be preferable over `Thread.sleep()`?

Comment (jonmv) [reply container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:106]: Hmm, not sure, after all.

**Comment by jonmv** [comment]
Yes, this is nice.

**Review by hmusum — APPROVED** [review]
LGTM

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java [hunk 2]
```diff
@@ -60,26 +61,57 @@ public void render(OutputStream output, ContentChannel networkChannel, Completio
     }
 
 
-    private static class BlockingFlushContentChannelOutputStream extends ContentChannelOutputStream {
-
+    private static class MaxPendingContentChannelOutputStream extends ContentChannelOutputStream {
         private final ContentChannel channel;
```
Comment (jonmv) [inline container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:65]: Now unused.

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java [hunk 3]
```diff
@@ -60,26 +61,57 @@ public void render(OutputStream output, ContentChannel networkChannel, Completio
...
+        private final long maxPending;
+        private AtomicLong sent = new AtomicLong(0);
+        private AtomicLong acked = new AtomicLong(0);
```
Comment (jonmv) [inline container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:68]: Should be `final`.

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java [hunk 4]
```diff
@@ -60,26 +61,57 @@ public void render(OutputStream output, ContentChannel networkChannel, Completio
...
+            @Override
+            public void failed(Throwable t) {
+                acked.addAndGet(written);
+            }
+        }
+        @Override
+        public void send(ByteBuffer src) throws IOException {
+            try {
+                stallWhilePendingAbove(maxPending);
+            } catch (InterruptedException e) {}
```
Comment (jonmv) [inline container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:100]:
```suggestion
            } catch (InterruptedException e) {
                throw new IOException("Interrupted waiting for IO");
            }
```

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java [hunk 5]
```diff
@@ -60,26 +61,57 @@ public void render(OutputStream output, ContentChannel networkChannel, Completio
...
-                          new CompletionHandler() {
-                              @Override public void completed() { latch.countDown(); }
-                              @Override public void failed(Throwable t) { latch.countDown(); }
-                          });
             try {
-                latch.await();
+                stallWhilePendingAbove(0);
             }
             catch (InterruptedException e) {
                 throw new RuntimeException("Interrupted waiting for underlying IO to complete", e);
```
Comment (jonmv) [inline container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:117]:
```suggestion
                throw new IOException("Interrupted waiting for IO");
```

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java (hunk 1 above)
Comment (baldersheim) [reply container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:118]: If you use signalling you increase the cost as that is around 3x more costly than an atomic add operation. We do not need the immediate wakeup as the inflight amount is significantly larger than a 1ms window. In these cases I find it preferable to use sleep.

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java (hunk 3 above)
Comment (baldersheim) [reply container-core/src/main/java/com/yahoo/container/handler/LogHandler.java]: Fixed

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java (hunk 2 above)
Comment (baldersheim) [reply container-core/src/main/java/com/yahoo/container/handler/LogHandler.java]: Fixed

**Comment by baldersheim** [comment]
Take a good look at the exception handling as the tracking must be correct, if not unhappiness might occur.
This is significantly simpler in c++ with destructors...

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java (hunk 1 above)
Comment (jonmv) [reply container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:118]: Yup. Thanks for the numbers.

**Comment by jonmv** [comment]
You'd think an API which lets you submit an `onDone() + onError()` callback object wouldn't throw exceptions—glad you spotted this. Looks correct.

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java [hunk 6]
```diff
@@ -60,26 +61,57 @@ public void render(OutputStream output, ContentChannel networkChannel, Completio
...
 
-        public BlockingFlushContentChannelOutputStream(ContentChannel endpoint) {
+        public MaxPendingContentChannelOutputStream(ContentChannel endpoint, long maxPending) {
             super(endpoint);
             this.channel = endpoint;
+            this.maxPending = maxPending;
+        }
+
+        private long pendingBytes() {
+            return sent.get() - acked.get();
```
Comment (bjorncs) [inline container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:76]: Consider replacing with a single atomic that is decremented for `completed()`/`failed()`

**Review by bjorncs — APPROVED** [review]

File: container-core/src/main/java/com/yahoo/container/handler/LogHandler.java (hunk 6 above)
Comment (baldersheim) [reply container-core/src/main/java/com/yahoo/container/handler/LogHandler.java:76]: I have concluded that most of the time it is better to have separate counters both counting up.
1 - There will be less contention.
2 - You preserve more information both for debugging and if you want monitoring.

**Pull request merged**